tests/xdc/history_replication_dlq_test.go

```diff
@@ -0,0 +1,454 @@
+// THE SOFTWARE.
+
+package xdc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/suite"
+	"github.com/urfave/cli/v2"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	sdkclient "go.temporal.io/sdk/client"
+	sdkworker "go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+	"go.uber.org/fx"
+
+	enumspb "go.temporal.io/server/api/enums/v1"
+	replicationspb "go.temporal.io/server/api/replication/v1"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/replication"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/tests"
+	"go.temporal.io/server/tools/tdbg"
+)
+
+// This file contains tests for the history replication DLQ feature. It uses a faulty replication task executor to force
+// replication tasks to be added to the DLQ, and then uses the tdbg CLI to read the replication tasks from the DLQ and
+// verify that they are correct.
+
+type (
+	historyReplicationDLQSuite struct {
+		xdcBaseSuite
+		replicationTaskExecutorParams          replicationTaskExecutorParams
+		executionManagerParams                 executionManagerParams
+		namespaceReplicationTaskExecutorParams namespaceReplicationTaskExecutorParams
+		enableReplicationStream                bool
+	}
+	replicationTaskExecutorParams struct {
+		// We use an interface{} for these tasks because we don't care about the data, and the concrete type changes
+		// depending on whether we have streaming enabled for replication or not.
+		tasks chan interface{}
+	}
+	testReplicationTaskExecutor struct {
+		*replicationTaskExecutorParams
+		taskExecutor replication.TaskExecutor
+	}
+	executionManagerParams struct {
+		// This channel is sent to once we're done processing a request to add a message to the DLQ.
+		dlqRequests chan *persistence.PutReplicationTaskToDLQRequest
+	}
+	testExecutionManager struct {
+		*executionManagerParams
+		persistence.ExecutionManager
+	}
+	namespaceReplicationTaskExecutorParams struct {
+		// This channel is sent to once we're done processing a namespace replication task.
+		tasks chan *replicationspb.NamespaceTaskAttributes
+	}
+	testNamespaceReplicationTaskExecutor struct {
+		*namespaceReplicationTaskExecutorParams
+		replicationTaskExecutor namespace.ReplicationTaskExecutor
+	}
+	testExecutableTaskConverter struct {
+		*replicationTaskExecutorParams
+		converter replication.ExecutableTaskConverter
+	}
+	testExecutableTask struct {
+		*replicationTaskExecutorParams
+		replication.TrackableExecutableTask
+	}
+)
+
+const (
+	testTimeout      = 30 * time.Second
+	workflowIDToFail = "dlq-replication-test-workflow"
```

Comment:
nit: maybe put "failure" somewhere in the string so it's clearer when this appears in the test logs?

Reply:
I renamed this to "history-replication-dlq-test-workflow" to be clearer. I don't want to put "failure" in the name because the workflow itself succeeds; it's just its replication tasks that fail.